fl4health/losses/mkmmd_loss.py

```diff
@@ -0,0 +1,288 @@
+
+        #   This function solves a batch of QPs, each optimizing over
+        #   `nz` variables and having `nineq` inequality constraints
+        #   and `neq` equality constraints.
+        #   The optimization problem for each instance in the batch
+        #   (dropping indexing from the notation) is of the form
+
+        #     \hat z =   argmin_z 1/2 z^T Q z + p^T z
+        #             subject to Gz <= h
+        #                         Az  = b
+        # Dimensions are as follows:
+        #   Q is (kernel_num, kernel_num)
+        #   p is (kernel_num, 1)
+        #   G is (kernel_num, kernel_num)
+        #   h is (kernel_num, 1)
+        #   A is (1, kernel_num)
+        #   b is a scalar
+
+        # QPFunction returns betas in the shape 1 x num_kernels to we take the transpose for consistency
+        return QPFunction(verbose=False, solver=QPSolvers.CVXPY, check_Q_spd=False)(
+            regularized_Q_k, p, G, h, hat_d_per_kernel.t(), b
+        ).t()
+
+    def optimize_betas(self, X: torch.Tensor, Y: torch.Tensor, lambda_m: float = 1e-5) -> torch.Tensor:
+        # In this function, we assume that X, Y: n_samples, n_features
+        all_h_u_per_v_i = self.compute_all_h_u_per_v_i(X, Y)
+
+        # shape of hat_d_per_kernel is number of kernels x 1
+        hat_d_per_kernel = self.compute_hat_d_per_kernel(all_h_u_per_v_i)
+        # check to see that at least one of them is positive. If none of them are positive, then select a single kernel
+        # with largest hat_d, similar to the suggestion of Gretton et al. in "Optimal Kernel Choice for Large-Scale
+        # Two-Sample Tests", 2012
+
+        # shape of hat_Q_k is number of kernels x number of kernels
+        hat_Q_k = self.compute_hat_Q_k(all_h_u_per_v_i)
+        # Eigen shift hat_Q_k and scale by 2 as the QP setup scales by 1/2
+        regularized_Q_k = 2 * hat_Q_k + lambda_m * torch.eye(self.kernel_num).to(self.device)
+
+        if not torch.any(hat_d_per_kernel > 0):
+            log(INFO, f"None of the estimates for hat_d are positive: {hat_d_per_kernel.squeeze()}.")
+            return self.beta_with_extreme_kernel_base_values(
+                hat_d_per_kernel, regularized_Q_k, minimize_type_two_error=True
+            )
+
+        if self.minimize_type_two_error:
+            try:
+                unnormalized_betas = self.form_and_solve_qp(hat_d_per_kernel, regularized_Q_k)
+            except Exception as e:
+                # If we can't solve the QP due to infeasibility, then we keep the previous betas
+                log(INFO, f"{e} We keep previous betas.")
+                unnormalized_betas = self.betas
+        else:
+            # If we're trying to maximize the type II error, then we are trying to maximize a convex function over a
+            # convex polygon of beta values. So the maximum is found at one of the vertices
+            unnormalized_betas = self.get_best_vertex_for_objective_function(hat_d_per_kernel, regularized_Q_k)
```

Comment:
In our meeting, we briefly discussed clipping the betas at 0.0 (since we were sometimes seeing small negative betas in spite of the QP constraint.) We can do this to the unnormalized betas so they don't contribute oddly to the normalization 🙂